client/state/plugins/installed/actions.js

```diff
@@ -0,0 +1,360 @@
+		: siteHandler.wpcomPlugin( pluginId );
+
+	return pluginHandler;
+};
+
+export function activate( site, plugin ) {
+	return ( dispatch ) => {
+		const pluginId = getPluginId( site, plugin );
+		const defaultAction = {
+			action: ACTIVATE_PLUGIN,
+			siteId: site.ID,
+			pluginId,
+		};
+		dispatch( Object.assign( {}, defaultAction, { type: PLUGIN_ACTIVATE_REQUEST } ) );
+
+		const successCallback = ( data ) => {
+			dispatch( Object.assign( {}, defaultAction, { type: PLUGIN_ACTIVATE_REQUEST_SUCCESS, data } ) );
+		};
+
+		const errorCallback = ( error ) => {
+			// This error means it's already active.
+			if ( error && error.error === 'activation_error' ) {
+				successCallback( plugin );
+			}
+			dispatch( Object.assign( {}, defaultAction, { type: PLUGIN_ACTIVATE_REQUEST_FAILURE, error } ) );
+		};
+
+		return getPluginHandler( site, pluginId ).activate().then( successCallback ).catch( errorCallback );
+	};
+}
+
+export function deactivate( site, plugin ) {
+	return ( dispatch ) => {
+		const pluginId = getPluginId( site, plugin );
+		const defaultAction = {
+			action: DEACTIVATE_PLUGIN,
+			siteId: site.ID,
+			pluginId,
+		};
+		dispatch( Object.assign( {}, defaultAction, { type: PLUGIN_DEACTIVATE_REQUEST } ) );
+
+		const successCallback = ( data ) => {
+			dispatch( Object.assign( {}, defaultAction, { type: PLUGIN_DEACTIVATE_REQUEST_SUCCESS, data } ) );
+		};
+
+		const errorCallback = ( error ) => {
+			// This error means it's already inactive.
+			if ( error && error.error === 'deactivation_error' ) {
+				successCallback( plugin );
+			}
+			dispatch( Object.assign( {}, defaultAction, { type: PLUGIN_DEACTIVATE_REQUEST_FAILURE, error } ) );
+		};
+
+		return getPluginHandler( site, pluginId ).deactivate().then( successCallback ).catch( errorCallback );
+	};
+}
+
+export function update( site, plugin ) {
+	return ( dispatch ) => {
+		if ( ! plugin.update ) {
+			return Promise.reject( 'Error: Plugin already up-to-date.' );
+		}
+
+		const pluginId = getPluginId( site, plugin );
+		const defaultAction = {
+			action: UPDATE_PLUGIN,
+			siteId: site.ID,
+			pluginId,
+		};
+		dispatch( Object.assign( {}, defaultAction, { type: PLUGIN_UPDATE_REQUEST } ) );
+
+		const successCallback = ( data ) => {
+			dispatch( Object.assign( {}, defaultAction, { type: PLUGIN_UPDATE_REQUEST_SUCCESS, data } ) );
+		};
+
+		const errorCallback = ( error ) => {
+			dispatch( Object.assign( {}, defaultAction, { type: PLUGIN_UPDATE_REQUEST_FAILURE, error } ) );
+		};
+
+		if ( ! site.canUpdateFiles ) {
+			const cantUpdateError = new Error( 'Error: Can\'t update files on the site' );
+			errorCallback( cantUpdateError );
+			return Promise.reject( cantUpdateError );
+		}
+
+		return getPluginHandler( site, pluginId ).updateVersion().then( successCallback ).catch( errorCallback );
+	};
+}
+
+export function enableAutoupdate( site, plugin ) {
+	return ( dispatch ) => {
+		if ( ! utils.userCan( 'manage_options', site ) || ! site.canAutoupdateFiles ) {
```

Comment:
It'd be nice if this were validated using the existing [`canCurrentUser` selector](https://github.com/Automattic/wp-calypso/blob/cefb6aa92d6f68b5dcbde0c92be7b5bc5aff0211/client/state/current-user/selectors.js#L91-L110) outside of the action creator.